stdlib/public/stubs/Stubs.cpp

```diff
@@ -445,11 +445,27 @@ __mulodi4(di_int a, di_int b, int* overflow)
 #endif
 
 #if defined(__CYGWIN__) || defined(_WIN32)
+  #define strncasecmp _strnicmp
+#endif
+
+static bool swift_stringIsSignalingNaN(const char *nptr) {
+  const std::size_t len = std::strlen(nptr);
+  return (len == 4 && strncasecmp(nptr, "snan", len) == 0) ||
+         (len == 5 && (strncasecmp(nptr, "+snan", len) == 0 ||
+                       strncasecmp(nptr, "-snan", len) == 0));
+}
+
+#if defined(__CYGWIN__) || defined(_WIN32)
 // Cygwin does not support uselocale(), but we can use the locale feature 
 // in stringstream object.
 template <typename T>
 static const char *_swift_stdlib_strtoX_clocale_impl(
     const char *nptr, T *outResult) {
+  if (swift_stringIsSignalingNaN(nptr)) {
+    *outResult = std::numeric_limits<T>::signaling_NaN();
+    return nptr + std::strlen(nptr);
+  }
+  
```

Comment:
There's no reason to use `strncasecmp` instead of `strcasecmp` when comparing to a literal; it makes your code more complex and doesn't add any safety (the literal necessarily has a trailing NUL, and `strcasecmp` will necessarily stop when it reaches it), and forces you to traverse the string twice. I would tend to write this function like so:
~~~~
if (nptr[0] == '+' || nptr[0] == '-')
  nptr++;
return (0 == strcasecmp(nptr, "snan"));
~~~~
(I would like it even more if the logic to strip a leading sign character were centralized somewhere, but this is acceptable).